mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestDerivates.java

```diff
@@ -0,0 +1,411 @@
+        return Response.ok()
+            .entity(mcrContent,
+                new Annotation[] { MCRParams.Factory
+                    .get(MCRParam.Factory.get(MCRContentAbstractWriter.PARAM_OBJECTTYPE, derid.getTypeId())) })
+            .lastModified(lastModified)
+            .build();
+    }
+
+    @PUT
+    @Consumes({ MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON + ";charset=UTF-8" })
+    @Operation(summary = "Creates or updates MCRDerivate with the body of this request",
+        tags = MCRRestUtils.TAG_MYCORE_DERIVATE,
+        responses = {
+            @ApiResponse(responseCode = "400",
+                content = { @Content(mediaType = MediaType.TEXT_PLAIN) },
+                description = "'Invalid body content' or 'MCRObjectID mismatch'"),
+            @ApiResponse(responseCode = "201", description = "MCRDerivate successfully created"),
+            @ApiResponse(responseCode = "204", description = "MCRDerivate successfully updated"),
+        })
+    @MCRRequireTransaction
+    @Path("/{" + PARAM_DERID + "}")
+    public Response updateDerivate(
+        @Parameter(example = "mir_derivate_00004711") @PathParam(PARAM_DERID) MCRObjectID derid,
+        @Parameter(required = true,
+            description = "MCRObject XML",
+            examples = @ExampleObject("<mycoreobject ID=\"{mcrid}\" ..>\n...\n</mycorobject>")) InputStream xmlSource)
+        throws IOException {
+        //check preconditions
+        try {
+            long lastModified = MCRXMLMetadataManager.instance().getLastModified(derid);
+            if (lastModified >= 0) {
+                Date lmDate = new Date(lastModified);
+                Optional<Response> cachedResponse = MCRRestUtils.getCachedResponse(request, lmDate);
+                if (cachedResponse.isPresent()) {
+                    return cachedResponse.get();
+                }
+            }
+        } catch (Exception e) {
+            //ignore errors as PUT is idempotent
+        }
+        boolean create = true;
+        if (MCRMetadataManager.exists(derid)) {
+            validateDerivateRelation(mcrId, derid);
+            create = false;
+        }
+        MCRStreamContent inputContent = new MCRStreamContent(xmlSource, null, MCRDerivate.ROOT_NAME);
+        MCRDerivate derivate;
+        try {
+            derivate = new MCRDerivate(inputContent.asXML());
+            derivate.validate();
+        } catch (JDOMException | SAXException | MCRException e) {
+            throw new BadRequestException(Response.status(Response.Status.BAD_REQUEST)
+                .entity("Invalid body content")
+                .build(), e);
+        }
+        if (!derid.equals(derivate.getId())) {
+            throw new BadRequestException(Response.status(Response.Status.BAD_REQUEST)
+                .entity("MCRObjectID mismatch")
+                .build());
+        }
+        try {
+            if (create) {
+                MCRMetadataManager.create(derivate);
```

Comment:
If this happens the Derivate has no FileSystem, i think it would be nice if it would be created like in doCreateDerivate.

Reply:
✅  good point